phac-nml/irida-next
# Analysis Exports: Backend - Accept multiple workflow IDs

## What does this PR do and why?
This PR adds new logic to accept multiple workflow execution IDs for any analysis export

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
1. Set up the ability to run workflow executions locally as described in [PR 507](https://github.com/phac-nml/irida-next/pull/507)
2. Run a couple workflow executions
3. Once completed, export it with:
```
user = USER YOU MADE WORKFLOW EXECUTION WITH
params_1 = {'export_type' => 'analysis', 'export_parameters' => {'ids' => [FIRST_WORKFLOW_EXECUTION_ID, SECOND_WORKFLOW_EXECUTION_ID], 'analysis_type' => 'user'}, 'email_notification' => true, 'name' => 'test export'}
DataExports::CreateService.new(user, params_1).execute
```
4. After 30s, download the export either through the UI or with (replace `localhost` with `localhost:3000` in the given URL):
```
data_export = DataExport.last
Rails.application.routes.url_helpers.rails_blob_url(data_export.file)
```
5. Ensure the top level directory now contains folders named with the `workflow_execution.id` and that they contain the expected files. Ensure both `manifest.txt` and `manifest.json` contain the accurate folder naming and structure
6. Repeat the above with workflow executions from `Automated Workflow Executions`, The workflows can be set up with the steps described in [PR585](https://github.com/phac-nml/irida-next/pull/585). 
The following will need to be changed in your params:
- Replace `"analysis_type" => "user"` with `"analysis_type" => "project"`
- Add `namespace_id => NAMESPACE_ID OF THE PROJECT YOU'RE EXPORTING FROM` to `params["export_parameters"]`
- Ensure you're using a user that is an `owner or maintainer` of the project. 
- Example params:
```
user = USER WITH CORRECT ACCESS
params_2 = {'export_type' => 'analysis', 'export_parameters' => {'ids' => [FIRST_WORKFLOW_EXECUTION_ID, SECOND_WORKFLOW_EXECUTION_ID], 'analysis_type' => 'project', 'namespace_id' => 'PROJECT.NAMESPACE_ID'}, 'email_notification' => true, 'name' => 'test export'}
DataExports::CreateService.new(user, params_2).execute
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
